src/main/java/com/epam/aidial/core/service/ApplicationService.java

```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
+                                 Function<HttpClientRequest, String> requestMapper,
+                                 BiFunction<HttpClientResponse, String, R> responseMapper) {
+        CompletableFuture<R> resultFuture = new CompletableFuture<>();
+        AtomicReference<HttpClientRequest> requestReference = new AtomicReference<>();
+        AtomicReference<HttpClientResponse> responseReference = new AtomicReference<>();
+
+        RequestOptions requestOptions = new RequestOptions()
+                .setMethod(method)
+                .setAbsoluteURI(controllerEndpoint + path)
+                .setIdleTimeout(controllerTimeout);
+
+        httpClient.request(requestOptions)
+                .compose(request -> {
+                    requestReference.set(request);
+                    String body = requestMapper.apply(request);
+                    return request.send((body == null) ? "" : body);
+                })
+                .compose(response -> {
+                    if (response.statusCode() != 200) {
+                        throw new IllegalStateException("Controller API error. Code: " + response.statusCode());
+                    }
+
+                    responseReference.set(response);
+                    return response.body();
+                })
+                .map(buffer -> {
+                    HttpClientResponse response = responseReference.get();
+                    String body = buffer.toString(StandardCharsets.UTF_8);
+                    return responseMapper.apply(response, body);
+                })
+                .onSuccess(resultFuture::complete)
+                .onFailure(resultFuture::completeExceptionally);
+
+        try {
+            return resultFuture.get(controllerTimeout, TimeUnit.MILLISECONDS);
+        } catch (Throwable e) {
+            HttpClientRequest request = requestReference.get();
+
+            if (request != null) {
+                request.reset();
+            }
+
+            if (e instanceof ExecutionException) {
+                e = e.getCause();
+            }
+
+            throw e;
+        }
+    }
+
+    private String deploymentLockKey(ResourceDescription resource) {
+        return BlobStorageUtil.toStoragePath(lockService.getPrefix(), "deployment:" + resource.getAbsoluteFilePath());
+    }
+
+    private String encodeTargetFolder(ResourceDescription resource, String id) {
+        String location = resource.getBucketLocation()
+                          + DEPLOYMENTS_NAME + PATH_SEPARATOR
+                          + id + PATH_SEPARATOR;
+
+        String name = encryptionService.encrypt(location);
+        return ResourceDescription.fromDecoded(ResourceType.FILE, name, location, null).getUrl();
+    }
+
+    public static boolean isActive(Application application) {
+        return application != null && application.getFunction() != null && isActive(application.getFunction().getStatus());
+    }
+
+    private static boolean isActive(Application.Function.Status status) {
```

Comment:
Let's move the method to the class Status

Reply:
done